File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int maxPathSum(TreeNode *root, int &maxSum) {
        if (root == NULL) return 0;
        int left = max(0, maxPathSum(root->left, maxSum));
        int right = max(0, maxPathSum(root->right, maxSum));
        maxSum = max(maxSum, root->val + left + right);
        return root->val + max(left, right);
    }
public:
    int maxPathSum(TreeNode* root) {
        int maxSum = root->val;
        maxPathSum(root, maxSum);
        return maxSum;
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxPathSum(root *TreeNode) int {
    maxxPathSum := root.Val
    calculateMaxPathSum(root, &maxxPathSum)
    return maxxPathSum
}

func calculateMaxPathSum(root *TreeNode, maxxPathSum *int) int {
    if root == nil {
        return 0
    }

    leftPathSum := max(0, calculateMaxPathSum(root.Left, maxxPathSum))
    rightPathSum := max(0, calculateMaxPathSum(root.Right, maxxPathSum))

    *maxxPathSum = max(*maxxPathSum, root.Val + leftPathSum + rightPathSum)

    return root.Val + max(0, leftPathSum, rightPathSum)
}